cabal-bundler/src/CabalBundler/OpenBSD.hs

```diff
@@ -0,0 +1,140 @@
+        P.UnitTypeBuiltin -> return []
+        P.UnitTypeLocal -> return []
+        _ -> do
+            rev <- case P.uSha256 unit of
+                Just _  -> do
+                    pkgInfo <- maybe (throwM $ UnknownPackageName cpkgname) return $
+                        M.lookup cpkgname meta
+                    relInfo <- maybe (throwM $ UnknownPackageVersion cpkgname cversion) return $
+                        M.lookup cversion $ I.piVersions pkgInfo
+
+                    return $ fromIntegral (I.riRevision relInfo)
+
+                Nothing -> case P.uType unit of
+                    P.UnitTypeLocal   -> return 0
+                    t                 -> throwM $ UnknownUnitType cpkgname t
+
+            return $ [Dep
+                { depPackageName = cpkgname
+                , depVersion     = cversion
+                , depRevision    = rev
+                }]
+
+data MetadataException
+    = UnknownPackageName C.PackageName
+    | UnknownExecutable String [P.UnitId]
+    | UnknownPackageVersion C.PackageName C.Version
+    | UnknownUnitType C.PackageName P.UnitType
+  deriving Show
+
+instance Exception MetadataException
+
+bfs :: Map P.UnitId P.Unit -> P.UnitId -> Peu r [P.Unit]
+bfs units unit0 = fmap concat $ do
+    uids <- either (throwM . PackageLoop) id $ TG.runG am $ \g -> do
+        v <- maybe (throwM $ MissingUnit unit0) return $
+            TG.gToVertex g unit0
+
+        return $ map (TG.gFromVertex g) $
+            -- nub and sort
+            reverse $ S.toList $ S.fromList $ concat $ TG.dfs g v
+
+    for uids $ \uid -> do
+        unit <- lookupUnit units uid
+        exes <- case M.toList (P.uComps unit) of
+            [(_, compinfo)] ->
+                collectExeDeps units (P.ciExeDeps compinfo)
+            _ -> do
+                putDebug $ "Unit with multiple components " ++ show uid
+                pure []
+        pure $ [unit] <> exes
+
+  where
+    am :: M.Map P.UnitId (S.Set P.UnitId)
+    am = fmap (foldMap P.ciLibDeps . P.uComps) units
+
+data PlanConstructionException
+    = PackageLoop [P.UnitId]
+    | MissingUnit P.UnitId
+  deriving Show
+
+instance Exception PlanConstructionException
+
+collectExeDeps :: M.Map P.UnitId P.Unit -> S.Set P.UnitId -> Peu r [P.Unit]
+collectExeDeps units = traverse check . S.toList where
+    check uid = lookupUnit units uid
+
+lookupUnit :: M.Map P.UnitId P.Unit -> P.UnitId -> Peu r P.Unit
+lookupUnit units uid =
+    maybe (throwM $ MissingUnit uid) return $ M.lookup uid units
+
+manifestLine :: Dep -> String
+manifestLine dep = 
+    let name = prettyShow $ depPackageName dep
+        ver = prettyShow $ depVersion dep
+        rev = show $ depRevision dep
+     in intercalate "\t" ["", name, ver, rev, "\\"]
+
+findExe :: String -> Map P.UnitId P.Unit -> [P.UnitId]
+findExe exeName units =
+    [ uid
+    | (uid, unit) <- M.toList units
+    , (P.CompNameExe e, _) <- M.toList (P.uComps unit)
+    , Just (P.LocalUnpackedPackage _) <- [P.uPkgSrc unit]
```

Comment:
Why is this required?

In my branch (adapting to recent changes in framework) to take `packageName` and filter based on it.

```haskell
findExe :: PackageName ->  String -> Map P.UnitId P.Unit -> [P.UnitId]
findExe pn exeName units =
    [ uid
    | (uid, unit) <- M.toList units
    , pkgName (toCabal (P.uPId unit)) == pn
    , (P.CompNameExe e, _) <- M.toList (P.uComps unit)
    , e == T.pack exeName
    ]
```

Reply:
You may recall that this code was written in a different context (cabal-bootstrap-gen). I may have missed a better way to do it as I moved over to cabal-bundler. Maybe point me at your branch so I can test it?

Reply:
I found https://github.com/phadej/cabal-extras/tree/pr-32-openbsd-ports-manifest, will use that.

Reply:
Yes, that the one to look at. I just forcepushed the update there though.